utils/retrieve-state/src/lib.rs

```diff
@@ -289,50 +337,369 @@ pub fn put_block_with_deploys(
+        peer.shutdown.send(()).unwrap();
+        peer.join_handle.await.unwrap();
+    }
+
+    Ok(0)
+}
+
+struct Peer {
+    address: SocketAddr,
+    sender: UnboundedSender<PeerMsg>,
+    join_handle: JoinHandle<()>,
+    shutdown: oneshot::Sender<()>,
+}
+
+async fn maybe_construct_peer(
+    base_send: UnboundedSender<Result<PeerMsg, PeerError>>,
+    address: SocketAddr,
+    client: &Client,
+    state_root_hash: Digest,
+) -> Result<Peer, PeerError> {
+    let url = format!("http://{}:{}/rpc", address.ip(), address.port());
+    let base_send = base_send.clone();
+    let client = client.clone();
+    let read_future = get_trie(
+        &client,
+        &url,
+        GetTrieParams {
+            trie_key: state_root_hash,
+        },
+    );
+    let read_or_timeout = tokio::time::timeout(Duration::from_secs(1), read_future).await;
+    match read_or_timeout {
+        Ok(Ok(_)) => Ok(spawn_peer(base_send, address, client)),
+        Ok(Err(error)) => {
+            return Err(PeerError::FailedRpc {
+                peer_address: address,
+                trie_key: state_root_hash,
+                error,
+            })
+        }
+        Err(elapsed) => {
+            return Err(PeerError::TimedOut {
+                peer_address: address,
+                trie_key: state_root_hash,
+                error: elapsed,
+            })
         }
-        if tries_downloaded % 1000 == 0 {
-            println!(
-                "downloaded {} tries in {}ms",
-                tries_downloaded,
-                start.elapsed().as_millis()
-            );
-            start = Instant::now();
+    }
+}
+
+/// Spawns a task to handle requests sent to this peer. This task will query the peer for a given
+/// Trie under a Digest. The returned Digest will be written to LMDB global state.
+fn spawn_peer(
+    base_send: UnboundedSender<Result<PeerMsg, PeerError>>,
+    address: SocketAddr,
+    client: Client,
+) -> Peer {
+    let (sender, peer_recv) = futures_channel::mpsc::unbounded::<PeerMsg>();
+    let peer_recv = peer_recv.map(|msg| Either::Left(msg));
+
+    let (shutdown, shutdown_recv) = futures_channel::oneshot::channel::<()>();
+    let shutdown_recv = shutdown_recv.into_stream().map(|_| Either::Right(()));
+
+    let mut next_peer_msg_or_shutdown_stream = futures::stream::select(peer_recv, shutdown_recv);
+
+    let join_handle = tokio::spawn(async move {
+        let client = client.clone();
+        'peer: while let Some(Either::Left(PeerMsg::GetTrie(next_trie_key))) =
+            next_peer_msg_or_shutdown_stream.next().await
+        {
+            let start = Instant::now();
+            let mut base_send = base_send.clone();
+            let url = format!("http://{}:{}/rpc", address.ip(), address.port());
+            match get_trie(
+                &client,
+                &url,
+                GetTrieParams {
+                    trie_key: next_trie_key,
+                },
+            )
+            .await
```

Comment:
If every peer is only a single task and always immediately `await`s the result of `get_trie` before making the next request, are we really ever making multiple requests to the same peer at once, as suggested by `concurrent_requests_per_peer`? Or are we just putting `concurrent_requests_per_peer` items into the channel at once?

Reply:
Correct, this is badly named and instead can be thought of as the `peer`'s mailbox in an actor model sense. I've renamed it appropriately.

Reply:
But then why do we have the `concurrent_requests_per_peer` setting at all?
I think we _do_ want to make that many requests at the same time, instead of `await`ing here immediately?

Reply:
Is `peer_mailbox_size` really useful? I think we'd really want to make concurrent requests instead.